MarkBind/markbind
# Support line numbers for code blocks

**What is the purpose of this pull request? (put "X" next to an item, remove the rest)**

• [ ] Documentation update
• [ ] Bug fix
• [ ] New feature
• [x] Enhancement to an existing feature
• [ ] Other, please explain:

Partially fixes #473 (only adding support for line nos),

Fixes #1023 





**What is the rationale for this request?**


**What changes did you make? (Give an overview)**

# Small bugfix: 

Our code fences (` ``` `) do not support [markdown-it-attrs](https://github.com/arve0/markdown-it-attrs) even though one of the examples there shows it should. 

Turns out this is due to our overriding of the syntax highlighting.


![image](https://user-images.githubusercontent.com/3646725/72735974-d2de1a80-3bd7-11ea-9e03-5b2ad462cb1e.png)

We copied the second example from markdown-it, so that we could add our own `hljs` class to the `<pre>` tag, but this breaks markdown-it-attrs since the attrs there do not get added to our hardcoded `pre` tag.

So instead we provide our own renderer to render the attributes parsed by markdown-it-attrs.

# Actual implementation

There is no way to directly use CSS and determine when line breaks happen, so we change `highlightLines` in markdown-it to split every `\n` and wrap a `span` around them.

Then, we use some css to insert the line numbers to it.

Line numbers will be shown by default, and authors can disable it by adding the class `no-line-numbers`.

**Provide some example code that this change will affect:**
````
```js
function f(n) {
  if (n < 1) {
    return 1;
  } else {
    return f(n - 1) + f(n - 2);
  }
}
f(10); // returns 12345;
```
````

Before and After:

![image](https://user-images.githubusercontent.com/3646725/72737875-636a2a00-3bdb-11ea-874f-6be23b40fe59.png)

If line numbers are not wanted:

````
```js {.no-line-numbers}
function f(n) {
  if (n < 1) {
    return 1;
  } else {
    return f(n - 1) + f(n - 2);
  }
}
f(10); // returns 12345;
```
````

**Is there anything you'd like reviewers to focus on?**

Check out the 2103 textbook here, where all code snippets have line numbers:
(the search on top doesn't work, it will bring you to the actual 2103 book site)
https://openorclose.github.io/se-book/

These pages have many code snippets:
- https://openorclose.github.io/se-book/codeQuality/
- https://openorclose.github.io/se-book/oop/
- https://openorclose.github.io/se-book/testing/
- https://openorclose.github.io/se-book/cppToJava/

**Testing instructions:**
Verify that `<span>`s get added to html of code fences, and manually check generated code fences for line numbers.

**Proposed commit message: (wrap lines at 72 characters)**

````
Add line numbers for code blocks

Let's wrap each line of a code block in a `<span>`,
and then add relevant css styles to display line numbers beside.

Line numbers will be displayed by default.

To turn off line numbers, add the class `no-line-numbers`
to a code block via markdown-it-attrs:

```js {.no-line-numbers}
const js = 1;
```
````
// TODO when design finalised
